core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1597,6 +1640,327 @@ public void addStateChangeListener(StateChangeListener<DistributedStagesSchedule
+                        coordinatorConsumedFragmentsBuilder.add(fragment.getId());
+                    }
+                    else {
+                        // create external exchange
+                        exchange = Optional.of(exchangeManager.create(new ExchangeContext(session.getQueryId(), stage.getStageId().getId()), hashPartitionCount));
+                        exchanges.put(fragment.getId(), exchange.get());
+                        taskLifecycleListener = TaskLifecycleListener.NO_OP;
+                    }
+
+                    ImmutableMap.Builder<PlanFragmentId, Exchange> sourceExchanges = ImmutableMap.builder();
+                    for (SqlStage childStage : stageManager.getChildren(fragment.getId())) {
+                        PlanFragmentId childFragmentId = childStage.getFragment().getId();
+                        Exchange sourceExchange = exchanges.get(childFragmentId);
+                        verify(sourceExchange != null, "exchange not found for fragment: %s", childFragmentId);
+                        sourceExchanges.put(childFragmentId, sourceExchange);
+                    }
+
+                    BucketToPartition inputBucketToPartition = bucketToPartitionMapCache.apply(fragment.getPartitioning());
+                    FaultTolerantStageScheduler scheduler = new FaultTolerantStageScheduler(
+                            session,
+                            stage,
+                            failureDetector,
+                            taskSourceFactory,
+                            nodeAllocator,
+                            taskLifecycleListener,
+                            exchange,
+                            bucketToPartitionMapCache.apply(fragment.getPartitioningScheme().getPartitioning().getHandle()).getBucketToPartitionMap(),
+                            sourceExchanges.build(),
+                            inputBucketToPartition.getBucketToPartitionMap(),
+                            inputBucketToPartition.getBucketNodeMap(),
+                            retryAttempts);
+
+                    schedulers.add(scheduler);
+                }
+
+                Set<PlanFragmentId> coordinatorConsumedFragments = coordinatorConsumedFragmentsBuilder.build();
+                stateMachine.addStateChangeListener(state -> {
+                    if (state == DistributedStagesSchedulerState.FINISHED) {
+                        coordinatorConsumedFragments.forEach(coordinatorTaskLifecycleListener::noMoreTasks);
+                    }
+                });
+
+                return new FaultTolerantDistributedStagesScheduler(
+                        stateMachine,
+                        queryStateMachine,
+                        schedulers.build(),
+                        schedulerStats,
+                        nodeAllocator,
+                        nodeUpdateTask);
+            }
+            catch (Throwable t) {
+                schedulers.build().forEach(FaultTolerantStageScheduler::abort);
```

Comment:
handle exceptions here.

It may be worthy to have a helper method which does the 
```
                catch (Throwable closeFailure) {
                    if (t != closeFailure) {
                        t.addSuppressed(closeFailure);
                    }
                }
```
part

Reply:
Added exception handling. Extracting `catch` into a method will require a callback that might decrease readability. Also the name and the interface will probably be weird (e.g.: `catchAndSuppress(Function, Throwable)`) that may potentially decrease readability.